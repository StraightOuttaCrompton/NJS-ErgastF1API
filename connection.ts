import mysql from "mysql";

const pool = mysql.createPool({
    connectionLimit: 20,
    host: "mysql-server",
    port: 3306,
    user: "root",
    password: "f1",
    database: "ergastdb",
});

export function getMySQLConnection() {
    return pool;
}

export function defaultLimit() {
    return 30;
}
