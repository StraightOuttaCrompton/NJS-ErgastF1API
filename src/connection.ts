import mysql from "mysql";

export const pool = mysql.createPool({
    connectionLimit: 20,
    host: process.env.NODE_ENV === "production" ? "mysql-server" : "localhost",
    port: 3306,
    user: "root",
    password: "f1",
    database: "ergastdb",
});

export function getMySQLConnection() {
    return pool;
}
