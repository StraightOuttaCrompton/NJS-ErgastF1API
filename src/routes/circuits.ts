import { DEFAULT_LIMIT, DEFAULT_OFFSET } from "../consts";
import { getMySQLConnection } from "../connection";
import { Request, Response } from "express";
import path from "path";

interface CircuitDB {
    circuitRef: string;
    url: string;
    name: string;
    lat: number;
    lng: number;
    alt: number;
    location: string;
    country: string;
}

interface Circuit {
    circuitId: string; // TODO: rename to id
    url: string;
    circuitName: string; // TODO: rename to name
    Location: {
        // TODO: rename to location
        lat: string; // TODO: number?
        long: string; // TODO: number?
        alt: string; // TODO: number?
        locality: string;
        country: string;
    };
}

function formatCircuits(rows: CircuitDB[]): Circuit[] {
    return rows.map((row) => ({
        circuitId: row.circuitRef,
        url: row.url,
        circuitName: row.name,
        Location: {
            lat: row.lat.toString(),
            long: row.lng.toString(),
            alt: row.alt != null ? row.alt.toString() : "N/D",
            locality: row.location,
            country: row.country,
        },
    }));
}

function parseParamAsInt(param: Request["query"][keyof Request["query"]], defaultValue: number) {
    if (typeof param === "string" || typeof param === "number") {
        const paramAsInt = parseInt(param);

        if (isNaN(paramAsInt)) {
            return defaultValue;
        }

        return paramAsInt;
    }

    return defaultValue;
}

interface CircuitResponse {
    MRData: {
        limit: string;
        offset: string;
        CircuitTable: {
            circuitId?: string;
            driverId?: string;
            constructorId?: string;
            grid?: string;
            result?: string;
            fastest?: string;
            status?: string;
            season?: string;
            round?: string;
            Circuits: Circuit[];
        };
    };
}

export function getCircuits(req: Request, res: Response) {
    const { offset: offsetParam, limit: limitParam } = req.query;

    const offset = parseParamAsInt(offsetParam, DEFAULT_OFFSET);
    const limit = parseParamAsInt(limitParam, DEFAULT_LIMIT);

    //START
    let year: any = null;
    let round: any = null;
    let constructor: any = null;
    let circuit: any = null;
    let driver: any = null;
    let grid: any = null;
    let result: any = null;
    let fastest: any = null;
    let status: any = null;
    let driverStandings: any = null;
    let constructorStandings: any = null;

    for (const key in req.query) {
        if (key != "offset" && key != "limit" && key != "sql") {
            switch (key) {
                case "year":
                    req.query[key] == "current"
                        ? (year = new Date().getFullYear().toString())
                        : (year = req.query[key]);
                    break;
                case "round":
                    round = req.query[key];
                    break;
                case "constructor":
                    constructor = req.query[key];
                    break;
                case "circuit":
                    circuit = req.query[key];
                    break;
                case "driver":
                    driver = req.query[key];
                    break;
                case "grid":
                    grid = req.query[key];
                    break;
                case "result":
                    result = req.query[key];
                    break;
                case "fastest":
                    fastest = req.query[key];
                    break;
                case "status":
                    status = req.query[key];
                    break;
                case "driverStandings":
                    driverStandings = req.query[key];
                    break;
                case "constructorStandings":
                    constructorStandings = req.query[key];
                    break;
                default:
                    res.status(400).send("Bad Request: Check the get params").end();
                    return;
                    break;
            }
        }
    }

    if (driverStandings || constructorStandings) {
        res.status(400).send("Bad Request: Circuit queries do not support standings qualifiers.").end();
        return;
    }

    let sql =
        "SELECT DISTINCT circuits.circuitRef, circuits.name, circuits.location, circuits.country, circuits.lat, circuits.lng, circuits.alt, circuits.url FROM circuits";
    if (year || driver || constructor || status || grid || fastest || result) sql += ", races";
    if (driver || constructor || status || grid || fastest || result) sql += ", results";
    if (driver) sql += ", drivers";
    if (constructor) sql += ", constructors";
    sql += " WHERE TRUE";

    //Set the join
    if (year || driver || constructor || status || grid || fastest || result)
        sql += " AND races.circuitId=circuits.circuitId";
    if (circuit) sql += ` AND circuits.circuitRef='${circuit}'`;
    if (driver || constructor || status || grid || fastest || result) sql += " AND results.raceId=races.raceId";
    if (constructor)
        sql += ` AND results.constructorId=constructors.constructorId AND constructors.constructorRef='${constructor}'`;
    if (driver) sql += ` AND results.driverId=drivers.driverId AND drivers.driverRef='${driver}'`;
    if (status) sql += ` AND results.statusId='${status}'`;
    if (grid) sql += ` AND results.grid='${grid}'`;
    if (fastest) sql += ` AND results.rank='${fastest}'`;
    if (result) sql += ` AND results.positionText='${result}'`;
    if (year) sql += ` AND races.year='${year}'`;
    if (round) sql += ` AND races.round='${round}'`;
    sql += ` ORDER BY circuits.circuitRef LIMIT ${offset}, ${limit}`;

    const mySQLConnection = getMySQLConnection();
    mySQLConnection.query(sql, (err, rows, fields) => {
        if (err) {
            console.log("Failed to query for " + __filename.slice(__filename.lastIndexOf(path.sep) + 1) + ": " + err);
            res.status(400).send({ error: err.sqlMessage, sql: err.sql }).end();
            return;
        }
        if (req.query.sql == "true") {
            res.status(200).send(sql).end();
            return;
        }

        const json: CircuitResponse = {
            MRData: {
                limit: limit.toString(),
                offset: offset.toString(),
                CircuitTable: {
                    Circuits: formatCircuits(rows),
                },
            },
        };

        if (circuit) json.MRData.CircuitTable.circuitId = circuit;
        if (driver) json.MRData.CircuitTable.driverId = driver;
        if (constructor) json.MRData.CircuitTable.constructorId = constructor;
        if (grid) json.MRData.CircuitTable.grid = grid;
        if (result) json.MRData.CircuitTable.result = result;
        if (fastest) json.MRData.CircuitTable.fastest = fastest;
        if (status) json.MRData.CircuitTable.status = status;
        if (year) json.MRData.CircuitTable.season = year;
        if (round) json.MRData.CircuitTable.round = round;

        res.json(json);
    });
}
