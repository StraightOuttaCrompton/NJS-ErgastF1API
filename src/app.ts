import express from "express";

import { getCircuits } from "./routes/circuits";
import routesDrivers from "./routes/drivers";
import routesConsturctors from "./routes/constructors";
import routesPitStop from "./routes/pitstops";
import routesLaps from "./routes/laps";
import routesStandingConstructors from "./routes/standingsConstructors";
import routesStandingDrivers from "./routes/standingsDrivers";
import routesRaceSchedule from "./routes/raceSchedule";
import routesStatus from "./routes/status";
import routesSeason from "./routes/season";
import routesQualifying from "./routes/qualifying";
import routesRaceResults from "./routes/raceResults";
import routesSprintResults from "./routes/sprintResults";

const app = express(); //TODO: make this configurable to limit access to known consumers?

//add Filters
app.get("/circuits", getCircuits);
app.use("/drivers", routesDrivers); //DONE
app.use("/constructors", routesConsturctors); //DONE
app.use("/pitstops", routesPitStop); //DONE
app.use("/laps", routesLaps); //DONE
app.use("/standings/constructors", routesStandingConstructors); //DONE
app.use("/standings/drivers", routesStandingDrivers); //DONE BUT THE PART OF CONSTRUCTORS IS MISSING
app.use("/races", routesRaceSchedule); //DONE
app.use("/status", routesStatus); //DONE
app.use("/seasons", routesSeason); //DONE
app.use("/qualifying", routesQualifying); //DONE
app.use("/results", routesRaceResults); //DONE
app.use("/results/sprint", routesSprintResults); //DONE

//TODO: add last and next params

app.get("", (req, res) => {
    res.status(200).send("Hi!");
});

app.use(function (req, res, next) {
    res.status(404).send("<h3>Bad Request</h3>");
});

export default app;
