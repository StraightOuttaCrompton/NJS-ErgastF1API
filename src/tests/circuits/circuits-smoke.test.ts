import request from "supertest";
import app from "../../app";
import querystring from "querystring";
import { pool } from "../../connection";
import { CircuitsQueryParameters } from "routes/circuits";

// import alonsoMclaren from "./expectedData/alonso-mclaren.json";
// import vettelWinsFerrari from "./expectedData/vettel-wins-ferrari.json";

describe("GET /circuits smoke tests", () => {
    const endpoint = "/circuits";

    const queries: Partial<CircuitsQueryParameters>[] = [
        {
            driverId: "bottas",
        },
        {
            driverId: "hamilton",
            year: 2022,
        },
        {
            driverId: "maldonado",
            status: 4,
        },
        {
            driverId: "hamilton",
            result: 2,
            year: 2022,
        },
        {
            driverId: "max_verstappen",
            grid: 1,
            year: 2022,
        },
    ];

    afterAll(async () => {
        pool.end();
    });

    const testCase = (url: string) => {
        test(url, async () => {
            const response = await request(app).get(url);

            expect(response.body).toMatchSnapshot();
        });
    };

    testCase(endpoint);

    queries.forEach((query) => {
        testCase(`${endpoint}?${querystring.stringify(query)}`);
    });

    // test("alonso-mclaren", async () => {
    //     const url = `${endpoint}?${querystring.stringify({
    //         driver: "alonso",
    //         constructor: "mclaren",
    //     })}`;

    //     const response = await request(app).get(url);

    //     expect(response.body).toEqual(alonsoMclaren);
    // });

    // test("vettel-wins-ferrari", async () => {
    //     const url = `${endpoint}?${querystring.stringify({
    //         driver: "vettel",
    //         constructor: "ferrari",
    //         result: 1,
    //     })}`;

    //     const response = await request(app).get(url);

    //     expect(response.body).toEqual(vettelWinsFerrari);
    // });
});
