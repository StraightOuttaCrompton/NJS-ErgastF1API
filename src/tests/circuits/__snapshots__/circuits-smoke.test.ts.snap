// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`circuit smoke tests all 1`] = `
[
  {
    "Location": {
      "alt": "58",
      "country": "Australia",
      "lat": "-34.9272",
      "locality": "Adelaide",
      "long": "138.617",
    },
    "circuitId": "adelaide",
    "circuitName": "Adelaide Street Circuit",
    "url": "http://en.wikipedia.org/wiki/Adelaide_Street_Circuit",
  },
  {
    "Location": {
      "alt": "19",
      "country": "Morocco",
      "lat": "33.5786",
      "locality": "Casablanca",
      "long": "-7.6875",
    },
    "circuitId": "ain-diab",
    "circuitName": "Ain Diab",
    "url": "http://en.wikipedia.org/wiki/Ain-Diab_Circuit",
  },
  {
    "Location": {
      "alt": "20",
      "country": "UK",
      "lat": "53.4769",
      "locality": "Liverpool",
      "long": "-2.94056",
    },
    "circuitId": "aintree",
    "circuitName": "Aintree",
    "url": "http://en.wikipedia.org/wiki/Aintree_Motor_Racing_Circuit",
  },
  {
    "Location": {
      "alt": "10",
      "country": "Australia",
      "lat": "-37.8497",
      "locality": "Melbourne",
      "long": "144.968",
    },
    "circuitId": "albert_park",
    "circuitName": "Albert Park Grand Prix Circuit",
    "url": "http://en.wikipedia.org/wiki/Melbourne_Grand_Prix_Circuit",
  },
  {
    "Location": {
      "alt": "161",
      "country": "USA",
      "lat": "30.1328",
      "locality": "Austin",
      "long": "-97.6411",
    },
    "circuitId": "americas",
    "circuitName": "Circuit of the Americas",
    "url": "http://en.wikipedia.org/wiki/Circuit_of_the_Americas",
  },
  {
    "Location": {
      "alt": "153",
      "country": "Sweden",
      "lat": "57.2653",
      "locality": "Anderstorp",
      "long": "13.6042",
    },
    "circuitId": "anderstorp",
    "circuitName": "Scandinavian Raceway",
    "url": "http://en.wikipedia.org/wiki/Scandinavian_Raceway",
  },
  {
    "Location": {
      "alt": "53",
      "country": "Germany",
      "lat": "52.4806",
      "locality": "Berlin",
      "long": "13.2514",
    },
    "circuitId": "avus",
    "circuitName": "AVUS",
    "url": "http://en.wikipedia.org/wiki/AVUS",
  },
  {
    "Location": {
      "alt": "7",
      "country": "Bahrain",
      "lat": "26.0325",
      "locality": "Sakhir",
      "long": "50.5106",
    },
    "circuitId": "bahrain",
    "circuitName": "Bahrain International Circuit",
    "url": "http://en.wikipedia.org/wiki/Bahrain_International_Circuit",
  },
  {
    "Location": {
      "alt": "-7",
      "country": "Azerbaijan",
      "lat": "40.3725",
      "locality": "Baku",
      "long": "49.8533",
    },
    "circuitId": "baku",
    "circuitName": "Baku City Circuit",
    "url": "http://en.wikipedia.org/wiki/Baku_City_Circuit",
  },
  {
    "Location": {
      "alt": "28",
      "country": "Portugal",
      "lat": "41.1705",
      "locality": "Oporto",
      "long": "-8.67325",
    },
    "circuitId": "boavista",
    "circuitName": "Circuito da Boavista",
    "url": "http://en.wikipedia.org/wiki/Circuito_da_Boavista",
  },
  {
    "Location": {
      "alt": "145",
      "country": "UK",
      "lat": "51.3569",
      "locality": "Kent",
      "long": "0.263056",
    },
    "circuitId": "brands_hatch",
    "circuitName": "Brands Hatch",
    "url": "http://en.wikipedia.org/wiki/Brands_Hatch",
  },
  {
    "Location": {
      "alt": "551",
      "country": "Switzerland",
      "lat": "46.9589",
      "locality": "Bern",
      "long": "7.40194",
    },
    "circuitId": "bremgarten",
    "circuitName": "Circuit Bremgarten",
    "url": "http://en.wikipedia.org/wiki/Circuit_Bremgarten",
  },
  {
    "Location": {
      "alt": "194",
      "country": "India",
      "lat": "28.3487",
      "locality": "Uttar Pradesh",
      "long": "77.5331",
    },
    "circuitId": "buddh",
    "circuitName": "Buddh International Circuit",
    "url": "http://en.wikipedia.org/wiki/Buddh_International_Circuit",
  },
  {
    "Location": {
      "alt": "109",
      "country": "Spain",
      "lat": "41.57",
      "locality": "Montmeló",
      "long": "2.26111",
    },
    "circuitId": "catalunya",
    "circuitName": "Circuit de Barcelona-Catalunya",
    "url": "http://en.wikipedia.org/wiki/Circuit_de_Barcelona-Catalunya",
  },
  {
    "Location": {
      "alt": "790",
      "country": "France",
      "lat": "45.7472",
      "locality": "Clermont-Ferrand",
      "long": "3.03889",
    },
    "circuitId": "charade",
    "circuitName": "Charade Circuit",
    "url": "http://en.wikipedia.org/wiki/Charade_Circuit",
  },
  {
    "Location": {
      "alt": "139",
      "country": "USA",
      "lat": "32.7774",
      "locality": "Dallas",
      "long": "-96.7587",
    },
    "circuitId": "dallas",
    "circuitName": "Fair Park",
    "url": "http://en.wikipedia.org/wiki/Fair_Park",
  },
  {
    "Location": {
      "alt": "177",
      "country": "USA",
      "lat": "42.3298",
      "locality": "Detroit",
      "long": "-83.0401",
    },
    "circuitId": "detroit",
    "circuitName": "Detroit Street Circuit",
    "url": "http://en.wikipedia.org/wiki/Detroit_street_circuit",
  },
  {
    "Location": {
      "alt": "484",
      "country": "France",
      "lat": "47.3625",
      "locality": "Dijon",
      "long": "4.89913",
    },
    "circuitId": "dijon",
    "circuitName": "Dijon-Prenois",
    "url": "http://en.wikipedia.org/wiki/Dijon-Prenois",
  },
  {
    "Location": {
      "alt": "88",
      "country": "UK",
      "lat": "52.8306",
      "locality": "Castle Donington",
      "long": "-1.37528",
    },
    "circuitId": "donington",
    "circuitName": "Donington Park",
    "url": "http://en.wikipedia.org/wiki/Donington_Park",
  },
  {
    "Location": {
      "alt": "81",
      "country": "France",
      "lat": "49.3306",
      "locality": "Rouen",
      "long": "1.00458",
    },
    "circuitId": "essarts",
    "circuitName": "Rouen-Les-Essarts",
    "url": "http://en.wikipedia.org/wiki/Rouen-Les-Essarts",
  },
  {
    "Location": {
      "alt": "130",
      "country": "Portugal",
      "lat": "38.7506",
      "locality": "Estoril",
      "long": "-9.39417",
    },
    "circuitId": "estoril",
    "circuitName": "Autódromo do Estoril",
    "url": "http://en.wikipedia.org/wiki/Aut%C3%B3dromo_do_Estoril",
  },
  {
    "Location": {
      "alt": "583",
      "country": "Japan",
      "lat": "35.3717",
      "locality": "Oyama",
      "long": "138.927",
    },
    "circuitId": "fuji",
    "circuitName": "Fuji Speedway",
    "url": "http://en.wikipedia.org/wiki/Fuji_Speedway",
  },
  {
    "Location": {
      "alt": "8",
      "country": "Argentina",
      "lat": "-34.6943",
      "locality": "Buenos Aires",
      "long": "-58.4593",
    },
    "circuitId": "galvez",
    "circuitName": "Autódromo Juan y Oscar Gálvez",
    "url": "http://en.wikipedia.org/wiki/Aut%C3%B3dromo_Oscar_Alfredo_G%C3%A1lvez",
  },
  {
    "Location": {
      "alt": "15",
      "country": "South Africa",
      "lat": "-33.0486",
      "locality": "Eastern Cape Province",
      "long": "27.8736",
    },
    "circuitId": "george",
    "circuitName": "Prince George Circuit",
    "url": "http://en.wikipedia.org/wiki/Prince_George_Circuit",
  },
  {
    "Location": {
      "alt": "103",
      "country": "Germany",
      "lat": "49.3278",
      "locality": "Hockenheim",
      "long": "8.56583",
    },
    "circuitId": "hockenheimring",
    "circuitName": "Hockenheimring",
    "url": "http://en.wikipedia.org/wiki/Hockenheimring",
  },
  {
    "Location": {
      "alt": "264",
      "country": "Hungary",
      "lat": "47.5789",
      "locality": "Budapest",
      "long": "19.2486",
    },
    "circuitId": "hungaroring",
    "circuitName": "Hungaroring",
    "url": "http://en.wikipedia.org/wiki/Hungaroring",
  },
  {
    "Location": {
      "alt": "37",
      "country": "Italy",
      "lat": "44.3439",
      "locality": "Imola",
      "long": "11.7167",
    },
    "circuitId": "imola",
    "circuitName": "Autodromo Enzo e Dino Ferrari",
    "url": "http://en.wikipedia.org/wiki/Autodromo_Enzo_e_Dino_Ferrari",
  },
  {
    "Location": {
      "alt": "223",
      "country": "USA",
      "lat": "39.795",
      "locality": "Indianapolis",
      "long": "-86.2347",
    },
    "circuitId": "indianapolis",
    "circuitName": "Indianapolis Motor Speedway",
    "url": "http://en.wikipedia.org/wiki/Indianapolis_Motor_Speedway",
  },
  {
    "Location": {
      "alt": "785",
      "country": "Brazil",
      "lat": "-23.7036",
      "locality": "São Paulo",
      "long": "-46.6997",
    },
    "circuitId": "interlagos",
    "circuitName": "Autódromo José Carlos Pace",
    "url": "http://en.wikipedia.org/wiki/Aut%C3%B3dromo_Jos%C3%A9_Carlos_Pace",
  },
  {
    "Location": {
      "alt": "130",
      "country": "Turkey",
      "lat": "40.9517",
      "locality": "Istanbul",
      "long": "29.405",
    },
    "circuitId": "istanbul",
    "circuitName": "Istanbul Park",
    "url": "http://en.wikipedia.org/wiki/Istanbul_Park",
  },
]
`;

exports[`circuit smoke tests bottas 1`] = `
[
  {
    "Location": {
      "alt": "10",
      "country": "Australia",
      "lat": "-37.8497",
      "locality": "Melbourne",
      "long": "144.968",
    },
    "circuitId": "albert_park",
    "circuitName": "Albert Park Grand Prix Circuit",
    "url": "http://en.wikipedia.org/wiki/Melbourne_Grand_Prix_Circuit",
  },
  {
    "Location": {
      "alt": "161",
      "country": "USA",
      "lat": "30.1328",
      "locality": "Austin",
      "long": "-97.6411",
    },
    "circuitId": "americas",
    "circuitName": "Circuit of the Americas",
    "url": "http://en.wikipedia.org/wiki/Circuit_of_the_Americas",
  },
  {
    "Location": {
      "alt": "7",
      "country": "Bahrain",
      "lat": "26.0325",
      "locality": "Sakhir",
      "long": "50.5106",
    },
    "circuitId": "bahrain",
    "circuitName": "Bahrain International Circuit",
    "url": "http://en.wikipedia.org/wiki/Bahrain_International_Circuit",
  },
  {
    "Location": {
      "alt": "-7",
      "country": "Azerbaijan",
      "lat": "40.3725",
      "locality": "Baku",
      "long": "49.8533",
    },
    "circuitId": "baku",
    "circuitName": "Baku City Circuit",
    "url": "http://en.wikipedia.org/wiki/Baku_City_Circuit",
  },
  {
    "Location": {
      "alt": "194",
      "country": "India",
      "lat": "28.3487",
      "locality": "Uttar Pradesh",
      "long": "77.5331",
    },
    "circuitId": "buddh",
    "circuitName": "Buddh International Circuit",
    "url": "http://en.wikipedia.org/wiki/Buddh_International_Circuit",
  },
  {
    "Location": {
      "alt": "109",
      "country": "Spain",
      "lat": "41.57",
      "locality": "Montmeló",
      "long": "2.26111",
    },
    "circuitId": "catalunya",
    "circuitName": "Circuit de Barcelona-Catalunya",
    "url": "http://en.wikipedia.org/wiki/Circuit_de_Barcelona-Catalunya",
  },
  {
    "Location": {
      "alt": "103",
      "country": "Germany",
      "lat": "49.3278",
      "locality": "Hockenheim",
      "long": "8.56583",
    },
    "circuitId": "hockenheimring",
    "circuitName": "Hockenheimring",
    "url": "http://en.wikipedia.org/wiki/Hockenheimring",
  },
  {
    "Location": {
      "alt": "264",
      "country": "Hungary",
      "lat": "47.5789",
      "locality": "Budapest",
      "long": "19.2486",
    },
    "circuitId": "hungaroring",
    "circuitName": "Hungaroring",
    "url": "http://en.wikipedia.org/wiki/Hungaroring",
  },
  {
    "Location": {
      "alt": "37",
      "country": "Italy",
      "lat": "44.3439",
      "locality": "Imola",
      "long": "11.7167",
    },
    "circuitId": "imola",
    "circuitName": "Autodromo Enzo e Dino Ferrari",
    "url": "http://en.wikipedia.org/wiki/Autodromo_Enzo_e_Dino_Ferrari",
  },
  {
    "Location": {
      "alt": "785",
      "country": "Brazil",
      "lat": "-23.7036",
      "locality": "São Paulo",
      "long": "-46.6997",
    },
    "circuitId": "interlagos",
    "circuitName": "Autódromo José Carlos Pace",
    "url": "http://en.wikipedia.org/wiki/Aut%C3%B3dromo_Jos%C3%A9_Carlos_Pace",
  },
  {
    "Location": {
      "alt": "130",
      "country": "Turkey",
      "lat": "40.9517",
      "locality": "Istanbul",
      "long": "29.405",
    },
    "circuitId": "istanbul",
    "circuitName": "Istanbul Park",
    "url": "http://en.wikipedia.org/wiki/Istanbul_Park",
  },
  {
    "Location": {
      "alt": "15",
      "country": "Saudi Arabia",
      "lat": "21.6319",
      "locality": "Jeddah",
      "long": "39.1044",
    },
    "circuitId": "jeddah",
    "circuitName": "Jeddah Corniche Circuit",
    "url": "http://en.wikipedia.org/wiki/Jeddah_Street_Circuit",
  },
  {
    "Location": {
      "alt": "N/D",
      "country": "Qatar",
      "lat": "25.49",
      "locality": "Al Daayen",
      "long": "51.4542",
    },
    "circuitId": "losail",
    "circuitName": "Losail International Circuit",
    "url": "http://en.wikipedia.org/wiki/Losail_International_Circuit",
  },
  {
    "Location": {
      "alt": "18",
      "country": "Singapore",
      "lat": "1.2914",
      "locality": "Marina Bay",
      "long": "103.864",
    },
    "circuitId": "marina_bay",
    "circuitName": "Marina Bay Street Circuit",
    "url": "http://en.wikipedia.org/wiki/Marina_Bay_Street_Circuit",
  },
  {
    "Location": {
      "alt": "N/D",
      "country": "USA",
      "lat": "25.9581",
      "locality": "Miami",
      "long": "-80.2389",
    },
    "circuitId": "miami",
    "circuitName": "Miami International Autodrome",
    "url": "http://en.wikipedia.org/wiki/Miami_International_Autodrome",
  },
  {
    "Location": {
      "alt": "7",
      "country": "Monaco",
      "lat": "43.7347",
      "locality": "Monte-Carlo",
      "long": "7.42056",
    },
    "circuitId": "monaco",
    "circuitName": "Circuit de Monaco",
    "url": "http://en.wikipedia.org/wiki/Circuit_de_Monaco",
  },
  {
    "Location": {
      "alt": "162",
      "country": "Italy",
      "lat": "45.6156",
      "locality": "Monza",
      "long": "9.28111",
    },
    "circuitId": "monza",
    "circuitName": "Autodromo Nazionale di Monza",
    "url": "http://en.wikipedia.org/wiki/Autodromo_Nazionale_Monza",
  },
  {
    "Location": {
      "alt": "255",
      "country": "Italy",
      "lat": "43.9975",
      "locality": "Mugello",
      "long": "11.3719",
    },
    "circuitId": "mugello",
    "circuitName": "Autodromo Internazionale del Mugello",
    "url": "http://en.wikipedia.org/wiki/Mugello_Circuit",
  },
  {
    "Location": {
      "alt": "578",
      "country": "Germany",
      "lat": "50.3356",
      "locality": "Nürburg",
      "long": "6.9475",
    },
    "circuitId": "nurburgring",
    "circuitName": "Nürburgring",
    "url": "http://en.wikipedia.org/wiki/N%C3%BCrburgring",
  },
  {
    "Location": {
      "alt": "108",
      "country": "Portugal",
      "lat": "37.227",
      "locality": "Portimão",
      "long": "-8.6267",
    },
    "circuitId": "portimao",
    "circuitName": "Autódromo Internacional do Algarve",
    "url": "http://en.wikipedia.org/wiki/Algarve_International_Circuit",
  },
  {
    "Location": {
      "alt": "678",
      "country": "Austria",
      "lat": "47.2197",
      "locality": "Spielberg",
      "long": "14.7647",
    },
    "circuitId": "red_bull_ring",
    "circuitName": "Red Bull Ring",
    "url": "http://en.wikipedia.org/wiki/Red_Bull_Ring",
  },
  {
    "Location": {
      "alt": "432",
      "country": "France",
      "lat": "43.2506",
      "locality": "Le Castellet",
      "long": "5.79167",
    },
    "circuitId": "ricard",
    "circuitName": "Circuit Paul Ricard",
    "url": "http://en.wikipedia.org/wiki/Paul_Ricard_Circuit",
  },
  {
    "Location": {
      "alt": "2227",
      "country": "Mexico",
      "lat": "19.4042",
      "locality": "Mexico City",
      "long": "-99.0907",
    },
    "circuitId": "rodriguez",
    "circuitName": "Autódromo Hermanos Rodríguez",
    "url": "http://en.wikipedia.org/wiki/Aut%C3%B3dromo_Hermanos_Rodr%C3%ADguez",
  },
  {
    "Location": {
      "alt": "18",
      "country": "Malaysia",
      "lat": "2.76083",
      "locality": "Kuala Lumpur",
      "long": "101.738",
    },
    "circuitId": "sepang",
    "circuitName": "Sepang International Circuit",
    "url": "http://en.wikipedia.org/wiki/Sepang_International_Circuit",
  },
  {
    "Location": {
      "alt": "5",
      "country": "China",
      "lat": "31.3389",
      "locality": "Shanghai",
      "long": "121.22",
    },
    "circuitId": "shanghai",
    "circuitName": "Shanghai International Circuit",
    "url": "http://en.wikipedia.org/wiki/Shanghai_International_Circuit",
  },
  {
    "Location": {
      "alt": "153",
      "country": "UK",
      "lat": "52.0786",
      "locality": "Silverstone",
      "long": "-1.01694",
    },
    "circuitId": "silverstone",
    "circuitName": "Silverstone Circuit",
    "url": "http://en.wikipedia.org/wiki/Silverstone_Circuit",
  },
  {
    "Location": {
      "alt": "2",
      "country": "Russia",
      "lat": "43.4057",
      "locality": "Sochi",
      "long": "39.9578",
    },
    "circuitId": "sochi",
    "circuitName": "Sochi Autodrom",
    "url": "http://en.wikipedia.org/wiki/Sochi_Autodrom",
  },
  {
    "Location": {
      "alt": "401",
      "country": "Belgium",
      "lat": "50.4372",
      "locality": "Spa",
      "long": "5.97139",
    },
    "circuitId": "spa",
    "circuitName": "Circuit de Spa-Francorchamps",
    "url": "http://en.wikipedia.org/wiki/Circuit_de_Spa-Francorchamps",
  },
  {
    "Location": {
      "alt": "45",
      "country": "Japan",
      "lat": "34.8431",
      "locality": "Suzuka",
      "long": "136.541",
    },
    "circuitId": "suzuka",
    "circuitName": "Suzuka Circuit",
    "url": "http://en.wikipedia.org/wiki/Suzuka_Circuit",
  },
  {
    "Location": {
      "alt": "13",
      "country": "Canada",
      "lat": "45.5",
      "locality": "Montreal",
      "long": "-73.5228",
    },
    "circuitId": "villeneuve",
    "circuitName": "Circuit Gilles Villeneuve",
    "url": "http://en.wikipedia.org/wiki/Circuit_Gilles_Villeneuve",
  },
]
`;
